src/ar_gateway_middleware.erl

```diff
@@ -237,11 +253,122 @@ get_tx_from_hash_1(TXID) ->
 		Filename -> {ok, TXID, Filename}
 	end.
 
-serve_tx(Filename, Req) ->
+serve_tx(Filename, Path, Fallback, Req, Env) ->
 	TX = ar_storage:read_tx_file(Filename),
-	ContentType = proplists:get_value(<<"Content-Type">>, TX#tx.tags, "text/html"),
-	Headers = #{<<"content-type">> => ContentType},
+	ContentType =
+		case lists:keyfind(<<"Content-Type">>, 1, TX#tx.tags) of
+			{<<"Content-Type">>, V} -> V;
+			false -> <<"text/html">>
+		end,
+	case {ContentType, Path} of
+		{?MANIFEST_CONTENT_TYPE, <<>>} ->
+			serve_manifest(TX, Req);
+		{?MANIFEST_CONTENT_TYPE, _} ->
+			serve_manifest_path(TX, Path, Fallback, Req, Env);
+		{_, <<>>} ->
+			serve_plain_tx(TX, ContentType, Req);
+		_ ->
+			bad_request(Req)
+	end.
+
+serve_manifest(TX, Req) ->
+	case ar_serialize:json_decode(TX#tx.data, [return_maps]) of
+		{ok, #{ <<"index">> := #{ <<"path">> := Index } }} ->
+			serve_manifest_index(Index, Req);
+		{ok, #{ <<"paths">> := Paths }} ->
+			serve_manifest_listing(ar_util:encode(TX#tx.id), maps:keys(Paths), Req);
+		{error, _} ->
+			misdirected_request(Req)
+	end.
+
+serve_manifest_index(Index, Req) ->
+	Hostname = cowboy_req:host(Req),
+	Path = cowboy_req:path(Req),
+	NewPath = filename:join(Path, Index),
+	Location = <<
+		"https://",
+		Hostname/binary,
+		NewPath/binary
+	>>,
+	{stop, cowboy_req:reply(301, #{<<"location">> => Location}, Req)}.
+
+serve_manifest_listing(TXID, Hrefs, Req) ->
+	Body = [
+		<<"<!doctype html>">>,
+		<<"<html>">>,
+		<<"<head>">>,
+		<<"<base href=\"/">>, html_escape(TXID), <<"/\">">>,
+		<<"<title>Arweave Listing</title>">>,
+		<<"</head>">>,
+		<<"<body>">>,
+		<<"<ul>">>,
+		[ [
+			<<"<li>">>,
+			<<"<a href=\"">>, html_escape(Href), <<"\">">>,
+			html_escape(Href),
+			<<"</a>">>,
+			<<"</li>">>
+		] || Href <- Hrefs],
+		<<"</ul>">>,
+		<<"</body>">>,
+		<<"</html>">>
+	],
+	{stop, cowboy_req:reply(200, #{ <<"content-type">> => <<"text/html">> }, Body, Req)}.
+
+html_escape(Html) ->
+	lists:foldl(fun({Char, Escape}, NextHtml) ->
+		binary:replace(NextHtml, Char, Escape, [global])
+	end, Html, [
+		{<<"&">>, <<"&amp;">>},
+		{<<"<">>, <<"&lt;">>},
+		{<<">">>, <<"&gt;">>},
+		{<<"\"">>, <<"&quot;">>},
+		{<<"'">>, <<"&apos;">>}
```

Comment:
I suppose, we need to escape `"` in addition to it `'` (also usually recommended in the XSS prevention guides).

Reply:
Already done 😅
https://github.com/ArweaveTeam/arweave/blob/557cd2e9f64fe6af46ddd25598691b6524192356/src/ar_gateway_middleware.erl#L325